### PR TITLE
Add nativeIO for madvise on mapped page cache

### PR DESCRIPTION
**Pull request by dstuebe**
Changes:
* Add c build step for jni
* Add nativeIO madvise system call wrapper
* Advise page tables and headers as 'will use'
* Add option to advise blob pages as 'random'
* Use mapped pages for writing in virtual pages files

Local benchmarking shows little impact on IO performance from use of madvise - this is good as impact might have been negative. Need to test under memory pressure at high IO in production environment. It would be ideal to develop a more rigorous local test with memory pressure.

The write to mapped change is an unexpected benefit. Need to test carefully in production as we have gone back and forth on this as a best practice. It looks like the latest iteration is a clear win for mapped IO as the benchmark time is cut in half!

See https://stackoverflow.com/a/49500154/2136991 for packaging & loading the dynamic library

### CODE REVIEW

File: src/nativeIO/c/com_upserve_uppend_blobs_NativeIO.c [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        return;
+    }
+
+    size_t size = (size_t) (*env)->GetDirectBufferCapacity(env, buffer);
+    if (size <= 0) {
+        (*env)->ThrowNew(env, class_ioex, strerror(errno));
+        return;
+    }
+
+    int page = getpagesize();
```
Comment (nicholasnassar) [inline src/nativeIO/c/com_upserve_uppend_blobs_NativeIO.c]: If you use this, you should have `#include <unistd.h>` at the top of the file. One of the other headers pulls it in anyway, but you shouldn't depend on that.

File: src/main/java/com/upserve/uppend/blobs/NativeIO.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    public enum Advice {
+        Normal(0), Sequential(1), Random(2), WillNeed(3), DontNeed(4);
+        private final int value;
+        Advice(int val) {
+            this.value = val;
+        }
+    }
+
+    public native void madvise(MappedByteBuffer buffer, Advice advise) throws IOException;
```
[the remark this answers is not in the document]
Comment (jeffrey-a-meunier) [reply src/main/java/com/upserve/uppend/blobs/NativeIO.java]: ^^ well spotted

File: src/nativeIO/c/com_upserve_uppend_blobs_NativeIO.c [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    end = (end + page-1)&(~(page-1));
+    size = (end-start);
+
+    //printf("\nDO madvise: page=%d p=0x%lx 0x%lx size=0x%lx\n", page, p, start, size);
+
+    int result = madvise((void *) start, size, osAdvice);
+    if (result != 0) {
+        sprintf(exBuffer, "system madvice call failed: '%d'", result);
+        (*env)->ThrowNew(env, class_ioex, exBuffer);
+        return ;
```
Comment (jeffrey-a-meunier) [inline src/nativeIO/c/com_upserve_uppend_blobs_NativeIO.c]: unnecessary return statement

File: src/nativeIO/c/com_upserve_uppend_blobs_NativeIO.c [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    return;
```
Comment (jeffrey-a-meunier) [inline src/nativeIO/c/com_upserve_uppend_blobs_NativeIO.c]: unnecessary return statement

File: src/main/java/com/upserve/uppend/blobs/VirtualPageFile.java [hunk 5]
```diff
@@ -95,8 +97,20 @@ public Path getFilePath() {
     @Override
     public void close() throws IOException {
         if (!channel.isOpen()) return;
+
+        for (MappedByteBuffer buf: mappedByteBuffers){
+            if (Objects.nonNull(buf)) nativeIO.madvise(buf, NativeIO.Advice.DontNeed);
```
[the remark this answers is not in the document]
Comment (dstuebe) [reply src/main/java/com/upserve/uppend/blobs/VirtualPageFile.java]: I was having OOM problems while running tests.
DirectBuffers are not actually freed immediately - even when the GC frees the Java buffer, the underlying page cache is managed by the OS...
I will investigate whether this makes any difference.

[the remark this answers is not in the document]
Comment (dstuebe) [reply src/main/java/com/upserve/uppend/blobs/VirtualPageFile.java]: Fixed - not needed!

**Comment by dstuebe** [comment]
Got the dylib included in the Jar - but I can't get the build to or load to support multiple targets (linux & mac)
It loads automatically from the jar but only assuming the target platform is mac (dylib not so)
Jitpack build [fails](https://jitpack.io/com/github/upserve/uppend/5392a9b/build.log) - apparently they don't support native build - no gcc/clang?

**Comment by dstuebe** [comment]
Frustrated with integration for JNI - difficult to build multiple targets, jitpack does not support at all.

Investigating use of [JNR-FFI](https://github.com/jnr/jnr-ffi) but use of [constants](https://github.com/jnr/jnr-constants) is a bit kludgy investigating adding madvise values there.

**Comment by dstuebe** [comment]
Closing based on revised implementation using JNR-FFI in #109

**Pull request closed without merging**